gnovm/pkg/gnolang/machine.go

```diff
@@ -237,6 +268,69 @@ func (m *Machine) RunMemPackage(memPkg *std.MemPackage, save bool) (*PackageNode
 	return pn, pv
 }
 
+func filterDuplicates(fset *FileSet) {
+	defined := make(map[Name]struct{}, 128)
+	for _, f := range fset.Files {
+		// returns true if name already defined
+		addName := func(n Name) bool {
+			if _, ok := defined[n]; ok {
+				return true // i--
+			}
+			defined[n] = struct{}{}
+			return false
+		}
+		for i := 0; i < len(f.Decls); i++ {
+			d := f.Decls[i]
+			var name Name
+			switch d := d.(type) {
+			case *FuncDecl:
+				name = d.Name
+				if d.IsMethod {
+					name = Name(derefStar(d.Recv.Type).String()) + "." + name
+				}
+			case *TypeDecl:
+				name = d.Name
+			case *ValueDecl:
+				// simpler, most common case
+				if len(d.NameExprs) == 1 {
+					name = d.NameExprs[0].Name
+					break
+				}
+
+				// more complex: filter out the NameExprs which are not
+				// redeclarations, and set them in d.NameExprs
+				newNames := make([]NameExpr, 0, len(d.NameExprs))
+				for _, nx := range d.NameExprs {
+					if !addName(nx.Name) {
+						newNames = append(newNames, nx)
+					}
+				}
+				// all new names are re-declarations, remove
+				if len(newNames) == 0 {
+					f.Decls = append(f.Decls[:i], f.Decls[i+1:]...)
+					i--
+					continue
+				}
+				d.NameExprs = newNames
+				continue
+			default:
+				continue
+			}
+			if addName(name) {
+				f.Decls = append(f.Decls[:i], f.Decls[i+1:]...)
+				i--
+			}
+		}
+	}
+}
+
+func derefStar(x Expr) Expr {
```

Comment:
gnovm/pkg/gnolang/helpers.go defines Deref() which does the opposite.

Maybe this should be defined in gnovm/pkg/gnolang/helpers.go and be called Destar() or Enref().

Reply:
I don't know. helpers.go currently contains mostly functions for AST constructions, not for "deconstruction". Are you sure about that?

I would agree on keeping this here and calling it "destar" instead.